### PR TITLE
Multiple drivers

### PR DESCRIPTION
**Pull request by jedwards4b**
Implementation of multiple MCT drivers as an option for multi-instance simulations. If multi-instance is enabled, N drivers are run, each with one instance.

Also (changes not directly related to multi-driver):

- Changed interface of `check_lockedfiles` (check_lockedfiles.py) to take a case instead of a caseroot.
- Use `case.get_env` instead of `EnvBuild` in check_lockedfiles.py
- Changed `check_case` (case_submit.py) to not take a caseroot input.
- Cleaned up memleak testing in `_check_for_memleak` (system_tests_common.py)
- Fixed bad format in `build_xcpl_nml` (buildnml.py)

Test suite: scripts_regression_tests.py
Test baseline: NA
Test namelist changes: NA
Test status: bit for bit
Fixes: #1704 
Fixes: #1714 

User interface changes?: new --multi-driver option to create_newcase and _C# modifier to tests

Update gh-pages html (Y/N)?: Y

Code review: @gold2718

### CODE REVIEW

File: scripts/lib/CIME/SystemTests/pre.py [hunk 1]
```diff
@@ -85,26 +85,40 @@ def run_phase(self): # pylint: disable=arguments-differ
...
+        multi_driver = self._case.get_value("MULTI_DRIVER")
+
         for comp in pause_comps:
+            if comp == "cpl":
+                if multi_driver:
+                    ninst = self._case.get_value("NINST_MAX")
+                else:
+                    ninst = 1
+            else:
+                ninst = self._case.get_value("NINST_{}".format(comp.upper()))
```
Comment (gold2718) [inline scripts/lib/CIME/SystemTests/pre.py]: Should there be a blank line here (end of `if`)?

File: scripts/lib/CIME/SystemTests/pre.py [hunk 2]
```diff
@@ -85,26 +85,40 @@ def run_phase(self): # pylint: disable=arguments-differ
...
+                # rfile2 has to match rfile1 (same time string)
+                parts = os.path.basename(rfile1).split(".")
+                glob_str = "*.{}".format(".".join(parts[len(parts)-4:]))
+                restart_files_2 = glob.glob(os.path.join(rundir2, glob_str))
+                expect((len(restart_files_2) == 1),
+                       "Missing case2 restart file, {}", glob_str)
+                rfile2 = restart_files_2[0]
+                ok = cprnc(comp, rfile1, rfile2, self._case, rundir2)[0]
+                logger.warning("CPRNC result for {}: {}".format(os.path.basename(rfile1), "PASS" if (ok == should_match) else "FAIL"))
+                compare_ok = compare_ok and (should_match == ok)
```
Comment (gold2718) [inline scripts/lib/CIME/SystemTests/pre.py]: I'm not sure what's wrong yet but both `./create_test PRE.f09_g17.ADESP.hobart_nag.drv-som` and `./create_test PRE_N3.f09_g17.ADESP.hobart_nag.drv-som` fail so I'm going to pause the review here until I get an idea of what's wrong.

File: config/config_tests.xml [hunk 3]
```diff
@@ -500,6 +500,17 @@ NODEFAIL          Tests restart upon detected node failure. Generates fake failu
...
+    <DESC>multi-driver validation vs single-instance (default length)</DESC>
+    <INFO_DBUG>1</INFO_DBUG>
+    <DOUT_S>FALSE</DOUT_S>
+    <CONTINUE_RUN>FALSE</CONTINUE_RUN>
+    <REST_OPTION>none</REST_OPTION>
+    <HIST_OPTION>$STOP_OPTION</HIST_OPTION>
+    <HIST_N>$STOP_N</HIST_N>
+    <MULTI_DRIVER>TRUE</MULTI_DRIVER>
+  </test>
+
```
Comment (jedwards4b) [inline config/config_tests.xml]: The comment on line 128 has nothing to do with this PR.  If it isn't correct please open an issue.   I think that adding more tests to the system can come after we accept this functionality PR and should also be a separate issue.

File: doc/source/users_guide/multi-instance.rst [hunk 4]
```diff
@@ -47,16 +45,21 @@ To run two instances of CAM, CLM, CICE, RTM and DOCN, invoke the following comma
...
+
+This configuration will run each component instance on the original 144 tasks but will generate two copies of the model (in the same executable) for a total of 288 tasks.
 
 3. Set up the case
 ::
 
    > ./case.setup
 
-A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called. 
+A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called.
```
Comment (jedwards4b) [inline doc/source/users_guide/multi-instance.rst]: that's not true  - it is correct as written.

File: doc/source/users_guide/multi-instance.rst [hunk 5]
```diff
@@ -47,16 +45,21 @@ To run two instances of CAM, CLM, CICE, RTM and DOCN, invoke the following comma
...
 When calling **case.setup** with the **env_mach_pes.xml** file specifically, these files are created in **$CASEROOT**:
 ::
```
Comment (jedwards4b) [inline doc/source/users_guide/multi-instance.rst]: I don't understand why you are commenting on code and comments that I haven't changed.   I think you are correct but are they really issues for this PR or should they be noted separately?

File: scripts/lib/CIME/SystemTests/pre.py (hunk 2 above)
Comment (jedwards4b) [inline scripts/lib/CIME/SystemTests/pre.py]: The issue is the ocn mask as was discussed earlier in the week wrt the master.  I don't think that this failure is unique to this branch is it?

Comment (jedwards4b) [inline scripts/lib/CIME/SystemTests/pre.py]: The correct tests are  PRE.f09_f09.ADESP.hobart_nag PRE.f09_f09.ADESP_TEST.hobart_nag and they both pass.

Comment (gold2718) [inline scripts/lib/CIME/SystemTests/pre.py]: Right. Also, it turns out the forcing file pointed to by drv-som is bad but that change was not in the commit message so I got confused. I'm still looking into things. Currently:
```
PRE_N3.f19_f19.ADESP.hobart_nag
```
passes but:
```
PRE_C3.f19_f19.ADESP.hobart_nag
```
is failing. Still have lots of other tests to do as well.

Comment (jedwards4b) [inline scripts/lib/CIME/SystemTests/pre.py]: Working on this one now.

Comment (jedwards4b) [inline scripts/lib/CIME/SystemTests/pre.py]: This test is working now.

**Comment by jedwards4b** [comment]
Rebased to master, @gold2718 awaiting your feedback.

File: scripts/lib/CIME/check_lockedfiles.py [hunk 6]
```diff
@@ -122,15 +123,13 @@ def check_lockedfiles(caseroot=None):
...
                     if "PIO_VERSION" in diffs.keys():
-                        f1obj.set_value("BUILD_STATUS", 2)
-                        f1obj.write()
+                        case.set_value("BUILD_STATUS", 2)
                         logging.critical("Changing PIO_VERSION requires running "
                                          "case.build --clean-all and rebuilding")
                     else:
-                        f1obj.set_value("BUILD_STATUS", 1)
-                        f1obj.write()
+                        case.set_value("BUILD_STATUS", 1)
```
Comment (gold2718) [inline scripts/lib/CIME/check_lockedfiles.py]: Are the changes in the module even related to multi_driver? If not, please open a new issue and submit a PR with these changes.

File: scripts/lib/CIME/buildnml.py [hunk 7]
```diff
@@ -78,7 +78,7 @@ def build_xcpl_nml(case, caseroot, compname):
         if ninst == 1:
             filename = os.path.join(rundir, "{}_in".format(compname))
         else:
-            filename = os.path.join(rundir, "{}_in_{:4.4d}".format(compname, i))
+            filename = os.path.join(rundir, "{}_in_{:04d}".format(compname, i))
```
Comment (gold2718) [inline scripts/lib/CIME/buildnml.py]: This seems like an unrelated bugfix which should be in its own PR.

File: scripts/lib/CIME/case_run.py [hunk 8]
```diff
@@ -244,7 +257,7 @@ def case_run(case, skip_pnl=False):
            "You are not calling the run script via the submit script. "
            "As a result, short-term archiving will not be called automatically."
            "Please submit your run using the submit script like so:"
-           " ./case.submit")
+           " ./case.submit Time: {}".format(get_timestamp()))
```
Comment (gold2718) [inline scripts/lib/CIME/case_run.py]: Is this change related to multi_driver? If not, please open an issue and submit it as a PR.

File: scripts/lib/CIME/get_timing.py [hunk 9]
```diff
@@ -150,11 +162,16 @@ def getTiming(self):
                 not continue_run:
             inittype = "TRUE"
 
-        binfilename = os.path.join(rundir, "timing", "model_timing_stats")
+        if inst > 0:
+            inst_label = '_%04d' % inst
```
Comment (gold2718) [inline scripts/lib/CIME/get_timing.py]: Shouldn't this use the `'_{:04d}'.format(inst)` syntax?

File: scripts/lib/CIME/get_timing.py [hunk 10]
```diff
@@ -150,11 +162,16 @@ def getTiming(self):
...
+        else:
+            inst_label = ''
+
+        binfilename = os.path.join(rundir, "timing", "model_timing%s_stats" % inst_label)
```
Comment (gold2718) [inline scripts/lib/CIME/get_timing.py]: Shouldn't this use the `"model_timing{}_stats".format(inst_label)` syntax?

File: src/build_scripts/buildlib.csm_share [hunk 11]
```diff
@@ -19,14 +19,27 @@ my $USE_ESMF_LIB	= `./xmlquery USE_ESMF_LIB	--value`;
...
+my $NINST_GLC		= 1;
+my $NINST_LND		= 1;
+my $NINST_OCN		= 1;
+my $NINST_ROF		= 1;
+my $NINST_WAV		= 1;
+my $NINST_ESP		= 1;
+
+if ($multi_driver eq "FALSE") {
+    $NINST_ATM		= `./xmlquery NINST_ATM		--value`;
+    $NINST_ICE		        = `./xmlquery NINST_ICE		--value`;
```
Comment (gold2718) [inline src/build_scripts/buildlib.csm_share]: Tab character sneaked through? EXTERMINATE!

File: src/components/data_comps/desp/desp_comp_mod.F90 [hunk 12]
```diff
@@ -219,8 +219,13 @@ subroutine desp_comp_init(EClock, espid, mpicom_in, phase, read_restart,    &
...
       !------------------------------------------------------------------------
 
-      iosystem => shr_pio_getiosys(trim(inst_name))
-      call shr_strdata_pioinit(SDESP, iosystem, shr_pio_getiotype(trim(inst_name)))
+      call shr_strdata_pioinit(SDESP, COMPID)
+
+      call mpi_comm_rank(global_comm, global_mype, ierr)
+      call mpi_comm_size(global_comm, global_numpes, ierr)
+
+      write(logunit,*)'DESP: I am global rank ',global_mype,' of ',global_numpes
```
Comment (gold2718) [inline src/components/data_comps/desp/desp_comp_mod.F90]: Is this write statement still necessary / useful?

File: src/components/xcpl_comps/xshare/dead_mod.F90 [hunk 13]
```diff
@@ -216,7 +216,7 @@ subroutine dead_setNewGrid(decomp_type,nxg,nyg,totpe,mype,lsize,gbuf,seg_len,npr
          i = i + 1
       enddo
 
-      write(logunit,*) 'dead_setNewGrid decomp seg ',mype,lsize,nx
+!      write(logunit,*) 'dead_setNewGrid decomp seg ',mype,lsize,nx
```
Comment (gold2718) [inline src/components/xcpl_comps/xshare/dead_mod.F90]: I agree, let's remove this.

File: src/drivers/mct/cime_config/config_component.xml [hunk 14]
```diff
@@ -1949,6 +1949,18 @@
     <desc>ROOTPE (mpi task in MPI_COMM_WORLD) for each component</desc>
   </entry>
 
+  <entry id="MULTI_DRIVER">
+    <type>logical</type>
+    <default_value>FALSE</default_value>
+    <valid_values>TRUE,FALSE</valid_values>
+    <group>mach_pes</group>
+    <file>env_mach_pes.xml</file>
+    <desc>MULTI_DRIVER mode provides a separate driver (and coupler) component for each
```
Comment (gold2718) [inline src/drivers/mct/cime_config/config_component.xml]: I would rephrase this line as:
<desc>MULTI_DRIVER mode provides a separate driver (and coupler component) for each

File: src/drivers/mct/main/cime_comp_mod.F90 [hunk 15]
```diff
@@ -843,7 +855,12 @@ subroutine cime_pre_init2()
    !| Initialize infodata
    !----------------------------------------------------------
 
-   call seq_infodata_init(infodata,nlfilename, GLOID, pioid)
+   if (len(cpl_inst_tag) > 0) then
```
Comment (gold2718) [inline src/drivers/mct/main/cime_comp_mod.F90]: Shouldn't this be `len_trim` instead of `len`?

File: src/drivers/mct/main/cime_comp_mod.F90 [hunk 16]
```diff
@@ -3573,7 +3587,7 @@ subroutine cime_run()
          call seq_rest_write(EClock_d, seq_SyncClock, infodata,       &
               atm, lnd, ice, ocn, rof, glc, wav, esp,                 &
               fractions_ax, fractions_lx, fractions_ix, fractions_ox, &
-              fractions_rx, fractions_gx, fractions_wx)
+              fractions_rx, fractions_gx, fractions_wx, tag=trim(cpl_inst_tag))
```
Comment (gold2718) [inline src/drivers/mct/main/cime_comp_mod.F90]: Why specify the name of a non-optional input? Seems inconsistent.

File: doc/source/users_guide/multi-instance.rst [hunk 17]
```diff
@@ -47,16 +45,21 @@ To run two instances of CAM, CLM, CICE, RTM and DOCN, invoke the following comma
...
+To use this mode change
+::
+   > ./xmlchange MULTI_DRIVER=TRUE
+
+This configuration will run each component instance on the original 144 tasks but will generate two copies of the model (in the same executable) for a total of 288 tasks.
 
 3. Set up the case
 ::
 
    > ./case.setup
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply doc/source/users_guide/multi-instance.rst]: works for me

File: doc/source/users_guide/multi-instance.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (jedwards4b) [reply doc/source/users_guide/multi-instance.rst]: Whether or not multi_driver does not change the result here.   I still maintain that this is correct as written.

File: doc/source/users_guide/multi-instance.rst [hunk 18]
```diff
@@ -79,17 +82,21 @@ Also, **case.setup** creates the following ``*_in_*`` files and ``*txt*`` files
    lnd_in_0001, lnd_in_0002
    rof_in_0001, rof_in_0002
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply doc/source/users_guide/multi-instance.rst]: Removed this erroneous statement.

File: scripts/Tools/check_case [hunk 19]
```diff
@@ -50,9 +50,8 @@ def _main_func(description):
 
     parse_command_line(sys.argv, description)
 
-    check_lockedfiles()
-
     with Case(read_only=False) as case:
+        check_lockedfiles(case)
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/Tools/check_case]: I cannot sort out at this point what is related to multi-driver and what is not.   This are problems that I found and fixed in the course of developing this branch.

File: scripts/lib/CIME/SystemTests/system_tests_common.py [hunk 20]
```diff
@@ -349,6 +352,24 @@ def compare_env_run(self, expected=None):
...
+        cpllogs = glob.glob(os.path.join(coupler_log_path, 'cpl*.log.*'))
+        lastcpllogs = []
+        if cpllogs:
+            lastcpllogs.append(max(cpllogs, key=os.path.getctime))
+            basename = os.path.basename(lastcpllogs[0])
+            suffix = basename.split('.',1)[1]
+            for log in cpllogs:
+                if log in lastcpllogs:
+                    continue
+                if log.endswith(suffix):
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/lib/CIME/SystemTests/system_tests_common.py]: space added

File: scripts/lib/CIME/XML/env_mach_pes.py [hunk 21]
```diff
@@ -20,6 +20,13 @@ def __init__(self, case_root=None, infile="env_mach_pes.xml", components=None):
         EnvBase.__init__(self, case_root, infile, schema=schema)
 
     def get_value(self, vid, attribute=None, resolved=True, subgroup=None, pes_per_node=None): # pylint: disable=arguments-differ
+
+        if vid == "NINST_MAX":
+            value = 1
+            for comp in self._components:
+                value = max(value, self.get_value("NINST_{}".format(comp)))
+            return value
+
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/lib/CIME/XML/env_mach_pes.py]: done

File: scripts/lib/CIME/check_lockedfiles.py (hunk 6 above)
Comment (jedwards4b) [reply scripts/lib/CIME/check_lockedfiles.py]: I don't remember and think that its a total waste of time to try to figure that out.

File: src/components/data_comps/desp/desp_comp_mod.F90 (hunk 12 above)
Comment (jedwards4b) [reply src/components/data_comps/desp/desp_comp_mod.F90]: Yes I think that it is and this code is only used in tests.

File: src/components/xcpl_comps/xshare/dead_mod.F90 (hunk 13 above)
Comment (jedwards4b) [reply src/components/xcpl_comps/xshare/dead_mod.F90]: done

File: src/drivers/mct/main/cime_comp_mod.F90 (hunk 16 above)
Comment (jedwards4b) [reply src/drivers/mct/main/cime_comp_mod.F90]: fixed

**Comment by jedwards4b** [comment]
I have made most of the requested changes.

**Comment by rljacob** [comment]
How about we call this "Ensemble drivers"?   "multiple drivers" to me is the work Mariana is doing for the NUOPC driver.

**Comment by jedwards4b** [comment]
Mutiple drivers is meant to distinguish from the multiple components mode.  I think that ensemble drivers is even more confusing.

File: doc/source/users_guide/multi-instance.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (gold2718) [reply doc/source/users_guide/multi-instance.rst]: I tried it and got different files depending on whether multi_driver was true or not.
Also, it cannot be correct as written as it refers to a single file but multiple files are created.

File: scripts/lib/CIME/check_lockedfiles.py (hunk 6 above)
Comment (gold2718) [reply scripts/lib/CIME/check_lockedfiles.py]: Well I do not understand why you have a double standard where it is okay to fix incidental _code_ issues found when working on a new feature but not okay to fix incidental _comment_ or _documentation_ issues found in the same process.
How about:

1. Fix the documentation/comment issues I pointed out
2. Include information about the incidental fixes in the PR text so that there is some record in the log that this was done.

and I will drop my request for proper separation between new features and bug fixes.

File: doc/source/users_guide/multi-instance.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (jedwards4b) [reply doc/source/users_guide/multi-instance.rst]: I just tried this again and got exactly the same result - perhaps if you are getting a different result you should document exactly what that difference is.

[the remark this answers is not in the document]
Comment (jedwards4b) [reply doc/source/users_guide/multi-instance.rst]:
```

[the remark this answers is not in the document]
Comment (jedwards4b) [reply doc/source/users_guide/multi-instance.rst]:
```
./xmlchange MULTI_DRIVER=TRUE
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) rm user_nl*
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ./case.setup --reset
Successfully cleaned batch script case.run
Setting resource.RLIMIT_STACK to -1 from (307200000, -1)
Creating batch script case.run
Writing case.run script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.case.run
Writing case.st_archive script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.st_archive
Locking file env_mach_pes.xml
Creating user_nl_xxx files for components and cpl
If an old case build already exists, might want to run 'case.build --clean' before building
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ls -ltr user*
-rw-r--r-- 1 jedwards ncar  848 Sep  1 11:28 user_nl_cpl
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:28 user_nl_cam_0001
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:28 user_nl_cam_0002
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:28 user_nl_clm_0001
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:28 user_nl_clm_0002
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:28 user_nl_cice_0001
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:28 user_nl_cice_0002
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:28 user_nl_docn_0001
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:28 user_nl_docn_0002
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:28 user_nl_mosart_0001
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:28 user_nl_mosart_0002
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:28 user_nl_cism_0001
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:28 user_nl_cism_0002
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) rm user_nl*
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ./xmlchange MULTI_DRIVER=False
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ./case.setup --reset
Successfully cleaned batch script case.run
Setting resource.RLIMIT_STACK to -1 from (307200000, -1)
Creating batch script case.run
Writing case.run script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.case.run
Writing case.st_archive script from input template /glade/p/work/jedwards/sandboxes/cesm2_0_alpha/cime.jet/config/cesm/machines/template.st_archive
Locking file env_mach_pes.xml
Creating user_nl_xxx files for components and cpl
If an old case build already exists, might want to run 'case.build --clean' before building
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
:) ls -ltr user*
-rw-r--r-- 1 jedwards ncar  848 Sep  1 11:29 user_nl_cpl
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:29 user_nl_cam_0001
-rw-r--r-- 1 jedwards ncar  115 Sep  1 11:29 user_nl_cam_0002
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:29 user_nl_clm_0001
-rw-r--r-- 1 jedwards ncar 1489 Sep  1 11:29 user_nl_clm_0002
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:29 user_nl_cice_0001
-rw-r--r-- 1 jedwards ncar  367 Sep  1 11:29 user_nl_cice_0002
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:29 user_nl_docn_0001
-rw-r--r-- 1 jedwards ncar  932 Sep  1 11:29 user_nl_docn_0002
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:29 user_nl_mosart_0001
-rw-r--r-- 1 jedwards ncar  595 Sep  1 11:29 user_nl_mosart_0002
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:29 user_nl_cism_0001
-rw-r--r-- 1 jedwards ncar  781 Sep  1 11:29 user_nl_cism_0002
cheyenne2: ~/sandboxes/cesm2_0_alpha/cime.jet/scripts/Fmulti
```

**Comment by jedwards4b** [comment]
@gold2718 I'm having trouble following the flow of this conversation, if you still fell that I have not addressed issues please be explicit about exactly what those are.

**Comment by gold2718** [comment]
```
: ./create_newcase --case noMulti --compset F2000_DEV --res ne30_ne30_mg17
: cd noMulti
: ./xmlchange NINST_ATM=2,NINST_ROF=2,NINST_OCN=2,NINST_ICE=2
: ./xmlchange MULTI_DRIVER=FALSE
: ./case.setup
: ./preview_namelists
: ./xmlquery NINST
        NINST: ['ATM:2', 'LND:1', 'ICE:2', 'OCN:2', 'ROF:2', 'GLC:1', 'WAV:1', 'ESP:1']
: ls Fmulti/user_nl_*
Fmulti/user_nl_cam        Fmulti/user_nl_cice_0002  Fmulti/user_nl_docn_0001
Fmulti/user_nl_cam_0001   Fmulti/user_nl_cism       Fmulti/user_nl_docn_0002
Fmulti/user_nl_cam_0002   Fmulti/user_nl_clm        Fmulti/user_nl_mosart_0001
Fmulti/user_nl_cice_0001  Fmulti/user_nl_cpl        Fmulti/user_nl_mosart_0002

```
Note that there are no `_NNNN` files for CISM, CLM, or CPL.
Now, do the exact same thing with multi_driver
```
: ./create_newcase --case Multi --compset F2000_DEV --res ne30_ne30_mg17
: cd Multi
: ./xmlchange NINST_ATM=2,NINST_ROF=2,NINST_OCN=2,NINST_ICE=2
: ./xmlchange MULTI_DRIVER=TRUE
: ./case.setup
: ./preview_namelists
: ./xmlquery NINST
        NINST: ['ATM:2', 'LND:1', 'ICE:2', 'OCN:2', 'ROF:2', 'GLC:1', 'WAV:1', 'ESP:1']
: ls Fmulti/user_nl_*
user_nl_cam        user_nl_cice_0002  user_nl_clm_0001   user_nl_docn_0002
user_nl_cam_0001   user_nl_cism_0001  user_nl_clm_0002   user_nl_mosart_0001
user_nl_cam_0002   user_nl_cism_0002  user_nl_cpl        user_nl_mosart_0002
user_nl_cice_0001  user_nl_clm        user_nl_docn_0001
```
So either the documentation is wrong or the `NINST` report is wrong or `xmlchange` is not working properly.

**Comment by jedwards4b** [comment]
I already updated the documentation in this PR to explicitly call out the additional changes to NINST:
```
   > ./xmlchange NINST_GLC=2                                                    
   > ./xmlchange NINST_LND=2            
```
so what are you looking at?

**Comment by gold2718** [comment]
I was not trying to follow the directions, I executed a series of commands and noted that the user_nl_xxx_NNNN files is not as described in multi-instance.rst.
I also wondered if the output is correct or if my command sequence should have generated an error.
Can you please respond to my post above?

**Comment by gold2718** [comment]
Another item which is not completed is that there is still no documentation block for the MCC test type in config_tests.xml.

**Comment by jedwards4b** [comment]
So the document provides directions that you didn't follow and now you don't understand why you got a different result?   I've updated config_tests.xml

**Comment by gold2718** [comment]
Yes, it is a new technique called 'testing'. Either the commands I entered are correct and the documentation (stating which user_nl files are created) is wrong, or the commands I entered are illegal and the code did not respond appropriately.
I tend to think I entered an illegal set of commands and the software should have either auto-corrected, warned or generated an error but I'm open to other interpretations.

File: doc/source/users_guide/multi-instance.rst [hunk 22]
```diff
@@ -1,95 +1,109 @@
...
+   'concurrent' for all but a few special cases and it cannot be used if MULTI_DRIVER=TRUE.
+
+#. In **create_test** these options can be invoked with testname modifiers _N# for the single driver mode and _C# for the multi-driver mode.  These are mutually exclusive options, they cannot be combined.
 
-#. Calling **case.setup** with ``--clean`` *DOES NOT* remove the **user_nl_xxx_NN** files created by **case.setup**.
+#. In create_newcase you may use --ninst # to set the number of instances and --multi-driver for multi-driver mode.
 
-#. Multiple instances generally should un concurrently, which is the default setting in **env_mach_pes.xml**. 
-   The serial setting is only for EXPERT USERS in upcoming development code implementations.
+#. In multi-driver mode you will always get 1 instance of each component for each driver/coupler, if you change a case using xmlchange MULTI_COUPLER=TRUE you will get a number of driver/couplers equal to the maximum NINST value over all components.
```
Comment (jedwards4b) [inline doc/source/users_guide/multi-instance.rst]: @gold2718 this note was added to the documentation to clarify use of multi-driver mode.   In my opinion there is no error in what you did to create a case, there is however an expectation that the user understand how multi-driver differs from multi-instance and thus would expect a different list of user_nl_XXXX files.

Comment (gold2718) [reply doc/source/users_guide/multi-instance.rst]: This seems confusion prone as `xmlquery NINST` does not show how many instances will actually run. @mvertens, @billsacks, any objections to this situation?

**Comment by gold2718** [comment]
Aside from the `xmlquery NINST` possible issue, the only thing I think is left is this unclear line in multi-instance.rst:

> A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instances) is generated when **case.setup** is called.

Can we please rephrase it to:

A new **user_nl_xxx_NNNN** file (where NNNN is the number of the component instance) is generated for each component instance when **case.setup** is called.

File: doc/source/users_guide/multi-instance.rst (hunk 22 above)
Comment (billsacks) [reply doc/source/users_guide/multi-instance.rst]: @gold2718 Sorry, I haven't been following this whole thread. I'd be happy to weigh in if you'd be able to give a brief summary of the situation that you want input on.

Comment (gold2718) [reply doc/source/users_guide/multi-instance.rst]: Look back about 6 posts to yesterday where I show two sets of commands (beginning with `./create_newcase`) and show that when `MULTI_DRIVER=TRUE`, the number of instances shown by `./xmlquery NINST` is not necessarily the number of instances which will be run.

Comment (billsacks) [reply doc/source/users_guide/multi-instance.rst]: Sorry, I took some time to try to understand the current behavior and expected behavior, but I'm still a bit lost. Could we have a brief in-person chat about this on Tuesday?

Comment (gold2718) [reply doc/source/users_guide/multi-instance.rst]: I don't want to take up a lot of your time with this but I will try a brief summary.
With a single driver, you can have N instances of some components but only one instance of others. For instance, you can have one ocean averaging the results of 64 atmospheres.

With multi-driver, each driver drives a separate instance of every non-stub component so if there are N drivers, there are N instances of every non-stub component.

The issue is that `xmlchange` allows you to set instance counts which are not viable with multi-driver but rather than display an error, the multi-driver code just ignores the situation and creates max(NINST_xxx) instances of every component. This behavior is mentioned at the end of the multi-instance documentation yet `xmlquery` still shows the values set by the user which is _not_ what is actually being run.

My question is whether this situation is okay or should be fixed. My concern is that users may read or hear that multi-driver speeds up the code so they turn it on. They can even check that they are running one ocean against 64 atmospheres (with `xmlquery`) but will in fact run 64 ocean instances with no averaging. As far as I can tell, there is no way to tell this is going to happen without running the case (or very carefully RTFM and we know how likely that is).

**Comment by mvertens** [comment]
I agree that an in-person chat would be useful. I'd like to be there as
well on Tuesday.

File: doc/source/users_guide/multi-instance.rst (hunk 22 above)
Comment (gold2718) [reply doc/source/users_guide/multi-instance.rst]: If we decide that there should be a fix, I propose the following actions:

1. When setting `MULTI_DRIVER` from `FALSE` to `TRUE`, check to make sure all `NINST_xxx` have the same value and generate an error if not.
2. If `MULTI_DRIVER == TRUE`, then `case.setup` will check that all `NINST_xxx` have the same value and generate an error if not.

**Comment by mvertens** [comment]
@gold2718, @jedwards4b -  I think this is a really good idea and one that
should be implemented to provide a robust implementation of this feature.

File: doc/source/users_guide/multi-instance.rst (hunk 22 above)
Comment (billsacks) [reply doc/source/users_guide/multi-instance.rst]: @gold2718 thanks for your explanation. Now I understand the issue. Yes, I agree with you that this basic sanity checking should be done, rather than silently running something different from what was requested and/or listed values would suggest.

**Comment by jedwards4b** [comment]
@billsacks @gold2718 I think that we need to sit down and talk about this when we can - it's not clear that we can put in the kind of error checking that you are requesting without creating new problems.  Suppose for example I add an error check that NINST_XXX has to be the same for all components in MULTI_DRIVER mode.   Now someone is in that mode and wants to change the NINST value from 3 to 4.   Should ```xmlchange NINST_ATM=4``` be an error (since all the others are still 3)?

**Comment by gold2718** [comment]
@jedwards4b,
I'm sick and not sure when I will be in next.
However, your example is why I did _not_ call for an NINST check in xmlchange. If you look at my two suggestions, they catch inconsistencies without (I think) leading to unwanted outcomes. Please review them (copied below for your convenience):

1. When setting MULTI_DRIVER from FALSE to TRUE, check to make sure all NINST_xxx have the same value and generate an error if not.
2. If MULTI_DRIVER == TRUE, then case.setup will check that all NINST_xxx have the same value and generate an error if not.

**Comment by jedwards4b** [comment]
1. Is easy, but I'm not sure it should be done because
2. leads to a situation wherein if I have MULTI_DRIVER=TRUE I cannot change the NINST_XXX values without generating an error.     We can talk when you are feeling better.

**Comment by gold2718** [comment]
I do not see how 2 leads to the situation. The only new xmlchange check is on setting MULTI_DRIVER from FALSE to TRUE. There is no consistency check when setting NINST_xxx.

**Comment by mvertens** [comment]
@gold2718 @jedwards @billsacks - at this point I think a quick face to face
meeting once @gold2718 feels better and is back will be the most productive
way to resolve this issue. I think this can be a short meeting where we
will be able reach consensus quickly on the best way forwards.

**Comment by jedwards4b** [comment]
@gold2718 I think I've figured out how to add the error check you requested.   Have another look.

**Comment by gold2718** [comment]
Feeling pretty good about this now.

**Comment by jedwards4b** [comment]
I've rerun scripts_regression_tests.py and fixed a pylint issue.  @billsacks are you still reviewing?

**Comment by billsacks** [comment]
I don't feel a need to review this.
